proposals/p2188.md

```diff
@@ -0,0 +1,517 @@
+    -   Pattern syntax makes it easier to match complex values.
+    -   Modeling pattern syntax after expressions eases the burden of learning a
+        new sub-language for pattern-matching: patterns are an extension of
+        expressions, and expressions are a special case of patterns.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   The rules for matching a templated value can be used to replace
+        `if constexpr` in many cases.
+
+## Alternatives considered
+
+### Struct pattern syntax
+
+We could omit the `, ...` syntax. This would simplify struct patterns, but at
+the cost of removing a feature that can be useful for reducing verbosity and
+making library evolution easier.
+
+We could always allow a struct pattern to match a struct with more fields,
+without requiring a `, ...` suffix. This would aid evolution by reducing the
+cases where adding a field to a struct can be a breaking change, but such cases
+would still exist. Further, this would make matches against a struct-valued
+expression inconsistent with matches against a struct pattern.
+
+We could remove the `{field: type}` shorthand and require
+`{.field = field: type}`. This may avoid encouraging reusing the field name even
+when it is not appropriate, and would remove some syntactic sugar that's not
+formally necessary. However, we expect this to be a common case whose ergonomics
+are important.
+
+We could use a different syntax for `{field: type}` that is less divergent from
+other struct syntaxes:
+
+-   `{.field: type}` seems like a contender, but doesn't work because that is
+    already recognized as a struct type literal. Also, the use of normal binding
+    syntax means that every locally-introduced name is always introduced by
+    `name: type` where `name` is not preceded by `.`.
+-   `{.=field: type}` might be a reasonable mnemonic shorthand for
+    `{.field = field: type}`, but looks a little surprising. This is probably
+    the best choice if concerns are found with `{field: type}` syntax.
+
+## Future work
+
+### Or patterns
+
+We could provide "or patterns", allowing matching of one pattern or another with
+the same handler:
+
+```
+match (x) {
+  case (m: i32, 0) | (0, m: i32) => { return m; }
+}
+```
+
+### Guards
+
+We could provide guards for patterns, allowing a pattern to match only if some
+predicate involving the bindings holds:
+
+```
+match (x) {
+  case (m: i32, n: i32) if m + n < 5 => { return m - n; }
+}
+```
+
+### Matching classes by dynamic type
+
+We could provide a way to match polymorphic class objects based on their dynamic
+type. This might be the default when matching a polymorphic class, or might
+require opt-in.
+
+### User-defined pattern matching
+
+We could provide some mechanism for allowing a user-defined type to specify how
+it can be matched by patterns.
```

Comment:
This is already more or less the plan of record.

```suggestion
We plan to provide a mechanism for allowing a user-defined type to specify how
it can be matched by patterns. See [p0157.md](p0157) for details.
```